### PR TITLE
Use static arrays for eigen-calculations

### PR DESCRIPTION
**Pull request by KnutAM**
Won't directly solve any issues, but rather move the responsibility of maintaining eigenvalue calculations to StaticArrays.jl
#173 and potentially #174/#175 (provided ForwardDiff [575](https://github.com/JuliaDiff/ForwardDiff.jl/pull/575)) could be solved by this. 
Gives less code base to maintain
Benchmarks to come soon...

### CODE REVIEW

**Comment by KnutAM** [comment]
Benchmark
```julia
using StableRNGs, BenchmarkTools, Random, Tensors
Base.rand(rng::AbstractRNG, ::Type{TT}) where{TT<:SymmetricTensor} = TT(ntuple((_)->rand(rng), Tensors.n_components(TT)))
rng=StableRNG(1);
for dim=1:3
        a=rand(rng,SymmetricTensor{2,dim})
        println("dim=$dim");
        @btime eigen($a);
        @btime eigvals($a);
        @btime eigvecs($a);
end
```

| dim |     eigen |   eigvals |   eigvecs | version |
| :----: | -------: | -------: | --------: | :------- |
| 1    |     0.800 |     0.900 |     0.800 | master |
| 1    |     1.000 |     0.800 |     1.000 | PR |
| 2    |     6.700 |     5.800 |     6.000 | master |
| 2    |     6.700 |     5.800 |     5.800 | PR |
| 3    | 488.205 | 488.205 | 488.718 | master |
| 3    | 513.542 | 513.021 | 515.625 | PR |

5.2% performance hit for 3-dimensional tensors. I'll update after using StaticArrays directly for eigvals and eigvecs, which might improve those cases without adding much code.

**Comment by KristofferC** [comment]
Seems like an OK performance hit to take.

File: Project.toml [hunk 1]
```diff
@@ -6,6 +6,7 @@ version = "1.10.0"
 ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 SIMD = "fdea26ae-647d-5447-a871-4b548cad5224"
+StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
```
Comment (fredrikekre) [inline Project.toml:9]: Needs an entry in `[compat]` below too. `StaticArrays = "1"` I guess.

File: src/eigen.jl [hunk 2]
```diff
@@ -1,242 +1,107 @@
-# MIT License: Copyright (c) 2016: Andy Ferris.
-# See LICENSE.md for further licensing test
+# Specify conversion to static arrays for 2nd order tensors
+@inline StaticArrays.SMatrix(a::Tensor{2,dim}) where{dim} = SMatrix{dim,dim}(a)
+@inline StaticArrays.SMatrix(a::SymmetricTensor{2,dim}) where{dim} = Symmetric(SMatrix{dim,dim}(a))
```
Comment (fredrikekre) [inline src/eigen.jl:3]: Could be internal functions instead
```suggestion
to_smatrix(a::Tensor{2, dim, T}) where {dim, T} = SMatrix{dim, dim, T}(a)
to_smatrix(a::SymmetricTensor{2, dim, T}) where {dim, T} = Symmetric(SMatrix{dim, dim, T}(a))
```

File: src/eigen.jl [hunk 3]
```diff
@@ -1,242 +1,107 @@
...
+@inline LinearAlgebra.eigvals(S::SymmetricTensor{4}) = eigvals(eigen(S))
+@inline LinearAlgebra.eigvals(S::SymmetricTensor{2,dim,T}) where{dim,T} = convert(Vec{dim,T}, Vec{dim}(eigvals(SMatrix(ustrip(S)))))
+
+"""
+    eigvecs(::SymmetricTensor)
+
+Compute the eigenvectors of a symmetric tensor.
+"""
+@inline LinearAlgebra.eigvecs(S::SymmetricTensor{4}) = eigvecs(eigen(S))
+@inline LinearAlgebra.eigvecs(S::SymmetricTensor{4}) = eigvecs(SMatrix(ustrip(S)))
```
Comment (fredrikekre) [inline src/eigen.jl:20]:
```suggestion
@inline LinearAlgebra.eigvecs(S::SymmetricTensor{4}) = eigvecs(to_smatrix(ustrip(S)))
```

File: src/eigen.jl [hunk 4]
```diff
@@ -1,242 +1,107 @@
...
-        Q[1, perm[3]], Q[2, perm[3]], Q[3, perm[3]],
-    ))
-    return Eigen(evals_sorted, Q_sorted)
+# Use specialized for 1d as this can be inlined
+@inline function LinearAlgebra.eigen(S::SymmetricTensor{2, 1, T}) where {T}
+    @inbounds Eigen(Vec{1, T}((S[1, 1],)), one(Tensor{2, 1, T}))
 end
+
+function LinearAlgebra.eigen(R::SymmetricTensor{2,dim,T}) where{dim,T}
+    e=eigen(SMatrix(ustrip(R)))
```
Comment (fredrikekre) [inline src/eigen.jl:105]:
```suggestion
    e = eigen(to_smatrix(ustrip(R)))
```

**Comment by KnutAM** [comment]
I can remove to_smatrix(::Tensor) (not used) and add test for eigvecs before it is ready for merge!
I'll look into that tomorrow!

**Comment by KnutAM** [comment]
Ready from my point of view now! 
Let me know if I should change something more!

**Pull request merged**